common/remote/rpc/rpc_client.go

```diff
@@ -0,0 +1,508 @@
+		}
+	}
+	if r.isShutdown() {
+		logger.Infof("%s Client is shutdown ,stop reconnect to server", r.name)
+	}
+}
+
+func (r *RpcClient) closeConnection() {
+	if r.currentConnection != nil {
+		r.currentConnection.close()
+		r.eventChan <- ConnectionEvent{eventType: DISCONNECTED}
+	}
+}
+
+// Notify when client new connected.
+func (r *RpcClient) notifyConnectionEvent(event ConnectionEvent) {
+	if len(r.connectionEventListeners) == 0 {
+		return
+	}
+	logger.Infof("%s Notify %s event to listeners.", r.name, event.toString())
+	for _, v := range r.connectionEventListeners {
+		if event.isConnected() {
+			v.OnConnected()
+		}
+		if event.isDisConnected() {
+			v.OnDisConnect()
+		}
+	}
+}
+
+func (r *RpcClient) healthCheck(timer *time.Timer) {
+	defer timer.Reset(constant.KEEP_ALIVE_TIME * time.Second)
+	var reconnectContext ReconnectContext
+	if time.Now().Sub(r.lastActiveTimeStamp) < constant.KEEP_ALIVE_TIME*time.Second {
+		return
+	}
+	if r.sendHealthCheck() {
+		r.lastActiveTimeStamp = time.Now()
+		return
+	} else {
+		if r.currentConnection == nil {
+			return
+		}
+		logger.Infof("%s Server healthy check fail,currentConnection=%s", r.name, r.currentConnection.getConnectionId())
+		atomic.StoreInt32((*int32)(&r.rpcClientStatus), (int32)(UNHEALTHY))
+		reconnectContext = ReconnectContext{onRequestFail: false}
+	}
+	r.reconnect(reconnectContext.serverInfo, reconnectContext.onRequestFail)
+}
+
+func (r *RpcClient) sendHealthCheck() bool {
+	if r.currentConnection == nil {
+		return false
+	}
+	response, err := r.currentConnection.request(rpc_request.NewHealthCheckRequest(),
+		constant.DEFAULT_TIMEOUT_MILLS, r)
+	if err != nil {
+		return false
+	}
+	if !response.IsSuccess() {
+		return false
+	}
+	return true
+}
+
+func (r *RpcClient) nextRpcServer() ServerInfo {
+	serverConfig := r.nacosServer.GetNextServer()
+	return ServerInfo{
+		serverIp:   serverConfig.IpAddr,
+		serverPort: serverConfig.Port,
+	}
+}
+
+func (c *ConnectionEvent) isConnected() bool {
+	return c.eventType == CONNECTED
+}
+
+func (c *ConnectionEvent) isDisConnected() bool {
+	return c.eventType == DISCONNECTED
+}
+
+//check is this client is shutdown.
+func (r *RpcClient) isShutdown() bool {
+	return atomic.LoadInt32((*int32)(&r.rpcClientStatus)) == (int32)(SHUTDOWN)
+}
+
+//check is this client is running.
+func (r *RpcClient) IsRunning() bool {
+	return atomic.LoadInt32((*int32)(&r.rpcClientStatus)) == (int32)(RUNNING)
+}
+
+func (c *ConnectionEvent) toString() string {
+	if c.isConnected() {
+		return "connected"
+	}
+	if c.isDisConnected() {
+		return "disconnected"
+	}
+	return ""
+}
+
+func (r *RpcClient) Request(request rpc_request.IRequest, timeoutMills int64) (rpc_response.IResponse, error) {
+	retryTimes := 0
+	start := util.CurrentMillis()
+	var currentErr error
+	for retryTimes < constant.REQUEST_DOMAIN_RETRY_TIME && util.CurrentMillis() < start+timeoutMills {
+		waitReconnect := false
+		if r.currentConnection == nil || !r.IsRunning() {
+			waitReconnect = true
```

Comment:
Is it unnecessary to add  "waitReconnect = true" , because it will return in next line.

Reply:
update